staging/src/k8s.io/apiextensions-apiserver/test/integration/validation_test.go

```diff
@@ -79,3 +80,166 @@ func TestForProperValidationErrors(t *testing.T) {
+	noxuVersionClient, err := testserver.CreateNewCustomResourceDefinition(noxuDefinition, apiExtensionClient, clientPool)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ns := "not-the-default"
+	noxuResourceClient := NewNamespacedCustomResourceClient(ns, noxuVersionClient, noxuDefinition)
+
+	tests := []struct {
+		name          string
+		instanceFn    func() *unstructured.Unstructured
+		expectedError string
+	}{
+		{
+			name: "bad alpha",
+			instanceFn: func() *unstructured.Unstructured {
+				instance := testserver.NewNoxuValidationInstance(ns, "foo")
+				instance.Object["spec"] = map[string]interface{}{
+					"alpha": "foo_123!",
+					"beta":  10,
+					"gamma": "bar",
+					"delta": "hello",
+				}
+				return instance
+			},
+			expectedError: "spec.alpha in body should match '^[a-zA-Z0-9_]*$'",
+		},
+		{
+			name: "bad beta",
+			instanceFn: func() *unstructured.Unstructured {
+				instance := testserver.NewNoxuValidationInstance(ns, "foo")
+				instance.Object["spec"] = map[string]interface{}{
+					"alpha": "foo_123",
+					"beta":  5,
+					"gamma": "bar",
+					"delta": "hello",
+				}
+				return instance
+			},
+			expectedError: "spec.beta in body should be greater than or equal to 10",
+		},
+		{
+			name: "bad gamma",
+			instanceFn: func() *unstructured.Unstructured {
+				instance := testserver.NewNoxuValidationInstance(ns, "foo")
+				instance.Object["spec"] = map[string]interface{}{
+					"alpha": "foo_123",
+					"beta":  5,
+					"gamma": "qux",
+					"delta": "hello",
+				}
+				return instance
+			},
+			expectedError: "spec.gamma in body should be one of [foo bar baz]",
+		},
+		{
+			name: "bad delta",
+			instanceFn: func() *unstructured.Unstructured {
+				instance := testserver.NewNoxuValidationInstance(ns, "foo")
+				instance.Object["spec"] = map[string]interface{}{
+					"alpha": "foo_123",
+					"beta":  5,
+					"gamma": "bar",
+					"delta": "foobarbaz",
+				}
+				return instance
+			},
+			expectedError: "must validate at least one schema (anyOf)\nspec.delta in body should be at most 5 chars long",
+		},
+		{
+			name: "absent alpha and beta",
+			instanceFn: func() *unstructured.Unstructured {
+				instance := testserver.NewNoxuValidationInstance(ns, "foo")
+				instance.Object["spec"] = map[string]interface{}{
+					"gamma": "bar",
+					"delta": "hello",
+				}
+				return instance
+			},
+			expectedError: "spec.alpha in body is required\nspec.beta in body is required",
+		},
+	}
+
+	for _, tc := range tests {
+		_, err := noxuResourceClient.Create(tc.instanceFn())
+		if err == nil {
+			t.Errorf("%v: expected %v", tc.name, tc.expectedError)
+			continue
+		}
+		// this only works when status errors contain the expect kind and version, so this effectively tests serializations too
+		if !strings.Contains(err.Error(), tc.expectedError) {
+			t.Errorf("%v: expected %v, got %v", tc.name, tc.expectedError, err)
+			continue
+		}
+	}
+}
```

Comment:
Can we have a test which updates a CRD validation and the checks that the new JSON schema is used?